kernel/smp.c

```diff
@@ -64,62 +103,128 @@ static void wait_for_start_signal(atomic_t *start_flag)
+	/* Wait for the CPU start caller to signal that
+	 * we can start initialization.
+	 */
+	wait_for_start_signal(&cpu_start_flag);
+
+	/* Initialize the dummy thread struct so that
+	 * the scheduler can schedule actual threads to run.
+	 */
 	z_dummy_thread_init(&dummy_thread);
+
 #ifdef CONFIG_SYS_CLOCK_EXISTS
-	smp_timer_init();
+	if ((csc == NULL) || ((csc != NULL) && csc->reinit_timer)) {
+		smp_timer_init();
+	}
 #endif
 
+	/* Do additional initialization steps if needed. */
+	if ((csc != NULL) && (csc->fn != NULL)) {
+		csc->fn(csc->arg);
+	}
+
+	/* Let scheduler decide what thread to run next. */
 	z_swap_unlocked();
 
 	CODE_UNREACHABLE; /* LCOV_EXCL_LINE */
 }
 
-static void start_cpu(int id, atomic_t *start_flag)
+static void start_cpu(int id, struct cpu_start_cb *csc)
 {
-	z_init_cpu(id);
+	/* Clear the ready flag so the newly powered up CPU can
+	 * signal that it has powered up.
+	 */
 	(void)atomic_clear(&ready_flag);
+
+	/* Power up the CPU */
 	arch_start_cpu(id, z_interrupt_stacks[id], CONFIG_ISR_STACK_SIZE,
-		       smp_init_top, start_flag);
+		       smp_init_top, (void *)csc);
+
+	/* Wait until the newly powered up CPU to signal that
+	 * it has powered up.
+	 */
 	while (!atomic_get(&ready_flag)) {
 		local_delay();
 	}
 }
 
-void z_smp_start_cpu(int id)
+void k_smp_cpu_start(int id, smp_init_fn fn, void *arg)
 {
+	k_spinlock_key_t key = k_spin_lock(&cpu_start_lock);
+
+	cpu_start_fn.fn = fn;
+	cpu_start_fn.arg = arg;
+
+#ifdef CONFIG_SYS_CLOCK_EXISTS
+	cpu_start_fn.reinit_timer = true;
+#endif
+
+	/* We are only starting one CPU so we do not need to synchronize
+	 * across all CPUs using the start_flag. So just set it to 1.
+	 */
 	(void)atomic_set(&cpu_start_flag, 1); /* async, don't care */
-	start_cpu(id, &cpu_start_flag);
+
+	/* Initialize various CPU structs related to this CPU. */
+	z_init_cpu(id);
+
+	/* Start the CPU! */
+	start_cpu(id, &cpu_start_fn);
+
+	k_spin_unlock(&cpu_start_lock, key);
+}
+
+void k_smp_cpu_resume(int id, smp_init_fn fn, void *arg,
+		      bool reinit_timer)
+{
+	k_spinlock_key_t key = k_spin_lock(&cpu_start_lock);
+
+	cpu_start_fn.fn = fn;
+	cpu_start_fn.arg = arg;
+
+#ifdef CONFIG_SYS_CLOCK_EXISTS
+	cpu_start_fn.reinit_timer = reinit_timer;
+#else
+	ARG_UNUSED(reinit_timer);
+#endif
+
+	/* We are only starting one CPU so we do not need to synchronize
+	 * across all CPUs using the start_flag. So just set it to 1.
+	 */
+	(void)atomic_set(&cpu_start_flag, 1); /* async, don't care */
+
+	/* Start the CPU! */
+	start_cpu(id, &cpu_start_fn);
+
+	k_spin_unlock(&cpu_start_lock, key);
 }
 
 void z_smp_init(void)
 {
+	/* We are powering up all CPUs and we want to synchronize their
+	 * entry into scheduler. So set the start flag to 0 here.
+	 */
 	(void)atomic_clear(&cpu_start_flag);
 
+	/* Just start CPUs one by one. */
 	unsigned int num_cpus = arch_num_cpus();
 
 	for (int i = 1; i < num_cpus; i++) {
-		start_cpu(i, &cpu_start_flag);
+		z_init_cpu(i);
+		start_cpu(i, NULL);
```

Comment:
Passing the NULL here requires you to add tests in a bunch of spaces in smp_init_top() to handle the NULL pointer.  Since the `cpu_start_fn` struct is static and available, can't you just set it up (a NULL function pointer and reinit_timer==true) to do what you want?

Reply:
This is done so that if `k_smp_cpu_start()` is not used at all, the garbage collector can remove the function and the static struct from the final binary... to save a few bytes.